### ISSUE TITLE
Submap for PGO?

### ISSUE THREAD

**Issue by YangSiri** [yanliang-wang/FAST_LIO_LC#6]
Hi, @yanliang-wang 

Did u use submap for loop closure or just single frame?

**Comment by yanliang-wang** [comment]
Hi, it is the submap form. 

If a loop closure is found, the registration result between their submap will be added to the pose graph.

**Comment by YangSiri** [comment]
Thanks, great work~

**Comment by YangSiri** [comment]
Hi, @yanliang-wang 

There are still some questions that i want to ask you for help. 
1) Did u notice the missing blocks in the scan? In my experiments with VLP16 and your gif as well, i saw some blocks of missing points in some rings of one scan. During the odometry, the positions of the missing block vary. Is this phenomenon caused by motion undistortion(certainly not by occlusion)? But still, i think it is weird in fastLIO.
![2022-05-10 10-23-28屏幕截图](https://user-images.githubusercontent.com/33154113/167529824-655d72ac-ca4f-439c-a470-987922f69b4b.png)
![2022-05-10 10-22-18屏幕截图](https://user-images.githubusercontent.com/33154113/167529872-8ed04b5e-b8b9-4442-b568-1da83412b560.png)

2) When i check the raw _time_ field in scan produced by _velodyne_driver_, it is negative. Do u know the reason for this?

**Comment by YangSiri** [comment]
Hi, @yanliang-wang 

I noticed you construct scancontext just using keframes instead of submap accumulated by kfs?

https://github.com/yanliang-wang/FAST_LIO_LC/blob/d05925567b3225fd0ffaa03d0098ee463299c5c4/PGO/src/laserPosegraphOptimization.cpp#L678

**Comment by yanliang-wang** [comment]
> Hi, @yanliang-wang
> 
> I noticed you construct scancontext just using keframes instead of submap accumulated by kfs?
> 
> https://github.com/yanliang-wang/FAST_LIO_LC/blob/d05925567b3225fd0ffaa03d0098ee463299c5c4/PGO/src/laserPosegraphOptimization.cpp#L678

Yes, this part is from [FAST_LIO_SLAM](https://github.com/gisbi-kim/FAST_LIO_SLAM). I didn't change it. But it doesn't matter if you only use the radius-based loop closure.

**Comment by YangSiri** [comment]
You were saying you also use frames to detect loop but match submaps to close the loop instead?

**Comment by yanliang-wang** [comment]
> You were saying you also use frames to detect loop but match submaps to close the loop instead?

Yes, is there any problem?

**Comment by YangSiri** [comment]
> Hi, it is the submap form.
> 
> If a loop closure is found, the registration result between their submap will be added to the pose graph.

Of course no, I misunderstood. I thought you use submap for loop detection as well. Thanks～
